### PR TITLE
bpo-31556: asyncio.wait_for can cancel futures faster with timeout <= 0

### PR DESCRIPTION
**Pull request by hellysmile**
https://bugs.python.org/issue31556

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept your contribution by verifying you have signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

Unfortunately our records indicate you have not signed the CLA. For legal reasons we need you to sign this before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.


Thanks again to your contribution and we look forward to looking at it!

File: Lib/asyncio/tasks.py [hunk 1]
```diff
@@ -334,6 +334,11 @@ def wait_for(fut, timeout, *, loop=None):
     if timeout is None:
         return (yield from fut)
 
+    if timeout == 0:
```
Comment (1st1) [inline Lib/asyncio/tasks.py]: You probably want to put `timeout <= 0` here to enable this behaviour for timeouts that were computed.

File: Misc/NEWS.d/next/Library/2017-09-22-23-48-49.bpo-31556.9J0u5H.rst [hunk 2]
```diff
@@ -0,0 +1 @@
+Cancel asyncio.wait_for future faster if timeout==0
```
Comment (1st1) [inline Misc/NEWS.d/next/Library/2017-09-22-23-48-49.bpo-31556.9J0u5H.rst]: Update the NEWS entry to reflect the '<=' change.

File: Lib/test/test_asyncio/test_tasks.py [hunk 3]
```diff
@@ -661,6 +661,37 @@ def task():
...
+            try:
+                yield from asyncio.sleep(0.2, loop=loop)
+            finally:
+                foo_running = False
+            return 'done'
+
+        fut = self.new_task(loop, foo())
+
+        with self.assertRaises(asyncio.TimeoutError):
+            loop.run_until_complete(asyncio.wait_for(fut, 0, loop=loop))
```
Comment (1st1) [inline Lib/test/test_asyncio/test_tasks.py]: Would be great if you cover both `timeout=0` and `timeout=-1` with `self.subTest`.

**Comment by 1st1** [comment]
LGTM. Please sign the CLA and I can merge this.

**Comment by hellysmile** [comment]
CLA was signed

File: Lib/asyncio/tasks.py [hunk 4]
```diff
@@ -334,6 +334,11 @@ def wait_for(fut, timeout, *, loop=None):
     if timeout is None:
         return (yield from fut)
 
+    if timeout <= 0:
+        fut = ensure_future(fut, loop=loop)
+        fut.cancel()
```
Comment (1st1) [inline Lib/asyncio/tasks.py]: Can you verify that the new code works the same way as the current asyncio code when `fut` is an already completed Future.

**Comment by hellysmile** [comment]
I think https://github.com/hellysmile/cpython/blob/9326df1bf94d781e3bf7d8b3140e6e3a405237b1/Lib/test/test_asyncio/test_tasks.py#L710 should be `self.assertIsNone(foo_running)` , but it fails give me a moment to check what is going on there

**Comment by hellysmile** [comment]
If future is created outside `wait_for` it has time for executing blocking code, but if coroutine object passed and future created inside `wait_for` is not started as expected, test for this case was added

**Comment by asvetlov** [comment]
Looks good.

**Comment by hellysmile** [comment]
It seems @the-knights-who-say-ni can not find my CLA, no green badge for 2 days. Is it makes any sense to try to sign it one more time?

**Comment by 1st1** [comment]
There's no `*` mark near your name on bugs.python.org.  Can you ping PSF?

**Comment by ncoghlan** [comment]
@1st1 Note that the signed CLAs are reviewed & recorded by PSF staff during US business hours, so it isn't unusual for signatures submitted on a Friday not to be noted until the following Monday. It's only when the next US business day has been and gone without the CLA signature being noted that the question may need to be escalated.

**Comment by 1st1** [comment]
@ncoghlan Oh, I didn't realize that I was suggesting to ping the PSF on Monday morning.. Sorry about that.

@hellysmile I'll need a bit more time to play with the PR before merging it, but so far I expect to merge it as is.

**Comment by hellysmile** [comment]
Python 3.7.0 alpha 2 comes 2017-10-16, is there any chance to include this patch to the upcoming alpha release?

**Review by 1st1 — APPROVED** [review]

**Comment by 1st1** [comment]
Merged. Thanks @hellysmile!

**Comment by hellysmile** [comment]
@1st1 Thank you!

**Pull request merged**